### PR TITLE
[WIP] Add identity detection to track facial identities over frames

### PR DESCRIPTION
**Pull request by ljchang**
This PR adds a new `identity_detector()` method that allows py-feat to detect embeddings of each detected face to address issue #164.

We use a pytorch implementation of the [facenet model](https://arxiv.org/pdf/1604.02878.pdf), which was trained to classify identities. The [pytorch version](https://github.com/timesler/facenet-pytorch) is a pytorch port of a [tensorflow implementation](https://github.com/davidsandberg/facenet), which was trained using the vggface2 dataset containing ~3.3M faces and ~9000 classes. It projects each face into a 512 dimensional latent embedding space. 

Here is an example, 

```python
# initialize py-feat detector instance
detector_cpu = Detector(verbose=False, device='cpu')

# load a test image
multi_face = os.path.join(get_test_data_path(), "multi_face.jpg")
frame = read_image(multi_face)

# detect face boxes
faces = detector_cpu.detect_faces(to_pil_image(frame))

# detect face identity embeddings
face_embeddings = detector_cpu.detect_identity(frame, faces)

# plot each detected face
face_imgs = [BBox(face[:-1]).extract_from_image(frame) for face in faces[0]]
f,a = plt.subplots(ncols=len(face_imgs))
for i,face in enumerate(face_imgs):
    a[i].imshow(to_pil_image(face))
    a[i].axis('off')
    a[i].set_title(f'Identity {i}')

# plot similarity of each detected face in embedding space
dist = 1 - pairwise_distances(face_embeddings.numpy(), metric='cosine')
sns.heatmap(dist)

```
Here are the detected faces
image

Here is their similarity in embedding space
image

You can see that two of the female faces are close in the embedding space where everyone else is more unique.  

We have a few things to figure out before we can finish this PR.

**1. What should `detect_identity()` output?**  currently it is just outputting the embeddings for each detected face, which I think makes the most sense. However, we could also output a prediction for each specific identity. Should the embedding output be a list of lists? or a tensor of all faces? or a hybrid list of tensors for each frame?  Currently it is a tensor of all faces.

**2. What should we include in our `Fex` output from our `_run_detection_waterfall()` method?**  I was thinking it would be both the embedding for each detected face and the most likely identity using a threshold keyword.  This would allow everyone to play with the embeddings themselves if they wanted to use different thresholds, but would also give a predicted identity.

**3. How should perform the inference on face identities?**
  - One possibility is to store a face dictionary inside the detector instance that can then be used as a lookup table for each subsequently detected face. the dictionary would grow as new identities are detected.
  - Another possibility is that we only perform the identity prediction when we are outputting the final full fex object in the detector_waterfall method. The advantage here is that it would be the most computationally efficient, but would need to be rerun if new faces were detected later.  I'm personally leaning towards this solution at the moment, as it would give us the most flexibility

### CODE REVIEW

**Comment by ljchang** [comment]
@ejolly. I ended up finishing adding the identity detector to py-feat.

Here are the choices I made:
1) `detect_identity()` will compute embeddings for each detected face and will make a prediction about face identity given the specified similarity `threshold` keyword argument. These will be output in a dictionary as a list of lists separately for predicted identity and the face embeddings.

2) identity predictions are made based on a rudimentary clustering algorithm based on cosine similarity using a user specified threshold implemented in `feat.utils.stats.cluster_identities()` function in pure pytorch, which should benefit from GPUs if they are available. Though this seems to already be very fast.

3) `detect images()` and `detect_video()` will now add identity detection to the full detection_waterfall pipeline

4) identity predictions can easily be recomputed on a Fex instance by running the `Fex.compute_identities()` method. It runs does not run in place by default, but can easily be changed. `updated_output = out.compute_identities(threshold=0.9)`

Let me know what you think and then all we need to is add tests for a couple of these functions.

**Comment by ljchang** [comment]
@ejolly: I've already discovered a weird bug, that I'm not sure how we will want to address.

Basically, if you have a list of images and are running a batchsize=1, it will detect different identities WITHIN a batch, not across batches.

```python
multi_face = os.path.join(get_test_data_path(), "multi_face.jpg")
single_face = os.path.join(get_test_data_path(), "single_face.jpg")
out = detector_cpu.detect_image([multi_face, single_face, multi_face], batch_size=1, face_identity_threshold=0.8)
out['Identity']
```
Screenshot 2023-09-15 at 3 36 03 PM

In this example., the identity_detector identifies 5 unique faces in the multi, but treats the new face in the single image separately, so it starts indexing at 0 again, finally in the repeat of the multi, it starts numbering over again.

However, if we were to compute the face predictions at the Fex construction after detection, it correctly identifies all of the unique faces and learns that Jin is a separate identity from the ones in the multi image.

```python
out_update = out.compute_identities(threshold=0.8, inplace=False)
out_update['Identity']
```
Screenshot 2023-09-15 at 3 35 22 PM

This would work correctly if all of the images were in the same batch, but this is making me rethink when we should compute the face prediction step.

**Comment by ljchang** [comment]
@ejolly, you were asking about how well the identity recognition was working and if it might be learning facial expression.  Just wanted to some quick results using our test video.

```python
video_path = os.path.join(get_test_data_path(), "WolfgangLanger_Pexels.mp4")
video_predictions = detector_cpu.detect_video(video_path, batch_size=10, skip_frames=24, face_identity_threshold=0.8)
```

here are different identities it learned with a threshold of 0.8 on the same person over time.

image

image

image

image

**Comment by ljchang** [comment]
@ejolly, i've updated the identity detection to now output list of lists like most of the other detectors and now identity prediction is done at the time of the fex construction. This ensures that identity prediction uses all of the faces in the fex instance instead of solely within the batch.

**Comment by ljchang** [comment]
@ejolly, i've updated the identity detection to now output list of lists like most of the other detectors and now identity prediction is done at the time of the fex construction. This ensures that identity prediction uses all of the faces in the fex instance instead of solely within the batch.

**Comment by ejolly** [comment]
@ljchang I've now disentangled this PR from #178 over at https://github.com/cosanlab/py-feat/pull/181 

This should be good to close without merging.
